File: cpp/fp2.cpp
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* CORE Fp^2 functions */
/* SU=m, m is Stack Usage (no lazy )*/

/* FP2 elements are of the form a+ib, where i is sqrt(-1) */

#include "fp2_YYY.h"

//namespace YYY {
//extern int fp2muls;
//extern int fp2sqrs;
//}

using namespace XXX;

/* test x==0 ? */
/* SU= 8 */
int YYY::FP2_iszilch(FP2 *x)
{
    return (FP_iszilch(&(x->a)) & FP_iszilch(&(x->b)));
}

/* Move g to f if d=1 */
void YYY::FP2_cmove(FP2 *f, FP2 *g, int d)
{
    FP_cmove(&(f->a), &(g->a), d);
    FP_cmove(&(f->b), &(g->b), d);
}

/* test x==1 ? */
/* SU= 48 */
int YYY::FP2_isunity(FP2 *x)
{
    FP one;
    FP_one(&one);
    return (FP_equals(&(x->a), &one) & FP_iszilch(&(x->b)));
}

/* SU= 8 */
/* Fully reduce a and b mod Modulus */
void YYY::FP2_reduce(FP2 *w)
{
    FP_reduce(&(w->a));
    FP_reduce(&(w->b));
}

/* return 1 if x==y, else 0 */
/* SU= 16 */
int YYY::FP2_equals(FP2 *x, FP2 *y)
{
    return (FP_equals(&(x->a), &(y->a)) & FP_equals(&(x->b), &(y->b)));
}

// Is x lexically larger than p-x?
// return -1 for no, 0 if x=0, 1 for yes
int YYY::FP2_islarger(FP2 *x)
{
    int cmp;
    if (FP2_iszilch(x)) return 0;
    cmp=FP_islarger(&(x->b));
    if (cmp!=0) return cmp;
    return FP_islarger(&(x->a));
}

void YYY::FP2_toBytes(char *b,FP2 *x)
{
    FP_toBytes(b,&(x->b));
    FP_toBytes(&b[MODBYTES_XXX],&(x->a));
}

void YYY::FP2_fromBytes(FP2 *x,char *b)
{
    FP_fromBytes(&(x->b),b);
    FP_fromBytes(&(x->a),&b[MODBYTES_XXX]);
}

/* Create FP2 from two FPs */
/* SU= 16 */
void YYY::FP2_from_FPs(FP2 *w, FP *x, FP *y)
{
    FP_copy(&(w->a), x);
    FP_copy(&(w->b), y);
}

/* Create FP2 from two BIGS */
/* SU= 16 */
void YYY::FP2_from_BIGs(FP2 *w, BIG x, BIG y)
{
    FP_nres(&(w->a), x);
    FP_nres(&(w->b), y);
}

/* Create FP2 from two ints */
void YYY::FP2_from_ints(FP2 *w, int xa, int xb)
{
    FP a,b;
    FP_from_int(&a,xa);
    FP_from_int(&b,xb);
    FP2_from_FPs(w,&a,&b);
}

/* Create FP2 from FP */
/* SU= 8 */
void YYY::FP2_from_FP(FP2 *w, FP *x)
{
    FP_copy(&(w->a), x);
    FP_zero(&(w->b));
}

/* Create FP2 from BIG */
/* SU= 8 */
void YYY::FP2_from_BIG(FP2 *w, BIG x)
{
    FP_nres(&(w->a), x);
    FP_zero(&(w->b));
}

/* FP2 copy w=x */
/* SU= 16 */
void YYY::FP2_copy(FP2 *w, FP2 *x)
{
    if (w == x) return;
    FP_copy(&(w->a), &(x->a));
    FP_copy(&(w->b), &(x->b));
}

/* FP2 set w=0 */
/* SU= 8 */
void YYY::FP2_zero(FP2 *w)
{
    FP_zero(&(w->a));
    FP_zero(&(w->b));
}

/* FP2 set w=1 */
/* SU= 48 */
void YYY::FP2_one(FP2 *w)
{
    FP one;
    FP_one(&one);
    FP2_from_FP(w, &one);
}

void YYY::FP2_rcopy(FP2 *w,const BIG a,const BIG b)
{
    FP_rcopy(&(w->a),a);
    FP_rcopy(&(w->b),b);
}

int YYY::FP2_sign(FP2 *w)
{
    int p1,p2;
    p1=FP_sign(&(w->a));
    p2=FP_sign(&(w->b));
#ifdef BIG_ENDIAN_SIGN_YYY
    p2 ^= (p1 ^ p2)&FP_iszilch(&(w->b));
    return p2;
#else
    p1 ^= (p1 ^ p2)&FP_iszilch(&(w->a));
    return p1;
#endif

}

/* Set w=-x */
/* SU= 88 */
void YYY::FP2_neg(FP2 *w, FP2 *x)
{
    /* Just one neg! */
    FP m, t;
    FP_add(&m, &(x->a), &(x->b));
    FP_neg(&m, &m);
    FP_add(&t, &m, &(x->b));
    FP_add(&(w->b), &m, &(x->a));
    FP_copy(&(w->a), &t);

}

/* Set w=conj(x) */
/* SU= 16 */
void YYY::FP2_conj(FP2 *w, FP2 *x)
{
    FP_copy(&(w->a), &(x->a));
    FP_neg(&(w->b), &(x->b));
    FP_norm(&(w->b));
}

/* Set w=x+y */
/* SU= 16 */
void YYY::FP2_add(FP2 *w, FP2 *x, FP2 *y)
{
    FP_add(&(w->a), &(x->a), &(y->a));
    FP_add(&(w->b), &(x->b), &(y->b));
}

/* Set w=x-y */
/* Input y MUST be normed */
void YYY::FP2_sub(FP2 *w, FP2 *x, FP2 *y)
{
    FP2 m;
    FP2_neg(&m, y);
    FP2_add(w, x, &m);
}

/* Set w=s*x, where s is FP */
/* SU= 16 */
void YYY::FP2_pmul(FP2 *w, FP2 *x, FP *s)
{
    FP_mul(&(w->a), &(x->a), s);
    FP_mul(&(w->b), &(x->b), s);
}

/* SU= 16 */
/* Set w=s*x, where s is int */
void YYY::FP2_imul(FP2 *w, FP2 *x, int s)
{
    FP_imul(&(w->a), &(x->a), s);
    FP_imul(&(w->b), &(x->b), s);
}

/* Set w=x^2 */
/* SU= 128 */
void YYY::FP2_sqr(FP2 *w, FP2 *x)
{
    FP w1, w3, mb;

    FP_add(&w1, &(x->a), &(x->b));
    FP_neg(&mb, &(x->b));

    FP_add(&w3, &(x->a), &(x->a));
    FP_norm(&w3);
    FP_mul(&(w->b), &w3, &(x->b));

    FP_add(&(w->a), &(x->a), &mb);

    FP_norm(&w1);
    FP_norm(&(w->a));

    FP_mul(&(w->a), &w1, &(w->a));   /* w->a#2 w->a=1 w1&w2=6 w1*w2=2 */

//  YYY::fp2sqrs++;
}


/* Set w=x*y */
/* Inputs MUST be normed  */
/* Now uses Lazy reduction */
void YYY::FP2_mul(FP2 *w, FP2 *x, FP2 *y)
{
    DBIG A, B, E, F, pR;
    BIG C, D, p;

    BIG_rcopy(p, Modulus);
    BIG_dsucopy(pR, p);

// reduce excesses of a and b as required (so product < pR)

    if ((sign64)(x->a.XES + x->b.XES) * (y->a.XES + y->b.XES) > (sign64)FEXCESS_YYY)
    {
#ifdef DEBUG_REDUCE
        printf("FP2 Product too large - reducing it\n");
#endif
        if (x->a.XES > 1) FP_reduce(&(x->a));
        if (x->b.XES > 1) FP_reduce(&(x->b));
    }

    BIG_mul(A, x->a.g, y->a.g);
    BIG_mul(B, x->b.g, y->b.g);

    BIG_add(C, x->a.g, x->b.g); BIG_norm(C);
    BIG_add(D, y->a.g, y->b.g); BIG_norm(D);

    BIG_mul(E, C, D);
    BIG_dadd(F, A, B);
    BIG_dsub(B, pR, B); //

    BIG_dadd(A, A, B);  // A<pR? Not necessarily, but <2pR
    BIG_dsub(E, E, F);  // E<pR ? Yes

    BIG_dnorm(A); FP_mod(w->a.g, A);  w->a.XES = 3; // may drift above 2p...
    BIG_dnorm(E); FP_mod(w->b.g, E);  w->b.XES = 2;

//  YYY::fp2muls++;
}

/* output FP2 in hex format [a,b] */
/* SU= 16 */
void YYY::FP2_output(FP2 *w)
{
    BIG bx, by;
    FP2_reduce(w);
    FP_redc(bx, &(w->a));
    FP_redc(by, &(w->b));
    printf("[");
    BIG_output(bx);
    printf(",");
    BIG_output(by);
    printf("]");
    FP_nres(&(w->a), bx);
    FP_nres(&(w->b), by);
}

/* SU= 8 */
void YYY::FP2_rawoutput(FP2 *w)
{
    printf("[");
    BIG_rawoutput(w->a.g);
    printf(",");
    BIG_rawoutput(w->b.g);
    printf("]");
}


/* Set w=1/x */
/* SU= 128 */
void YYY::FP2_inv(FP2 *w, FP2 *x, FP *h)
{
    FP w1, w2;

    FP2_norm(x);
    FP_sqr(&w1, &(x->a));
    FP_sqr(&w2, &(x->b));
    FP_add(&w1, &w1, &w2);
    FP_inv(&w1, &w1, h);
    FP_mul(&(w->a), &(x->a), &w1);
    FP_neg(&w1, &w1);
    FP_norm(&w1);
    FP_mul(&(w->b), &(x->b), &w1);
}


/* Set w=x/2 */
/* SU= 16 */
void YYY::FP2_div2(FP2 *w, FP2 *x)
{
    FP_div2(&(w->a), &(x->a));
    FP_div2(&(w->b), &(x->b));
}

/* Input MUST be normed */
void YYY::FP2_times_i(FP2 *w)
{
    FP z;
    FP_copy(&z, &(w->a));
    FP_neg(&(w->a), &(w->b));
    FP_copy(&(w->b), &z);

//    Output NOT normed, so use with care
}

/* Set w*=(2^i+sqrt(-1)) */
/* where X^2-(2^i+sqrt(-1)) is irreducible for FP4 */

/* Input MUST be normed */
void YYY::FP2_mul_ip(FP2 *w)
{
    FP2 t;
    int i = QNRI_YYY;

    FP2_copy(&t, w);
    FP2_times_i(w);

// add 2^i.t
    while (i > 0)
    {
        FP2_add(&t, &t, &t);
        FP2_norm(&t);
        i--;
    }
    FP2_add(w, &t, w);

#if TOWER_YYY == POSITOWER
    FP2_norm(w);
    FP2_neg(w, w);  // ***
#endif

//    Output NOT normed, so use with care
}

/* Set w/=(2^i+sqrt(-1)) */
/* Slow */
void YYY::FP2_div_ip(FP2 *w)
{
    FP2 z;
    FP2_norm(w);
    FP2_from_ints(&z, (1 << QNRI_YYY), 1);
    FP2_inv(&z, &z, NULL);
    FP2_mul(w, &z, w);
#if TOWER_YYY == POSITOWER
    FP2_neg(w, w);  // ***
#endif
}

/* SU= 8 */
/* normalise a and b components of w */
void YYY::FP2_norm(FP2 *w)
{
    FP_norm(&(w->a));
    FP_norm(&(w->b));
}

/* Set w=a^b mod m */
/* SU= 208 */
/*
void YYY::FP2_pow(FP2 *r, FP2* a, BIG b)
{
    FP2 w;
    FP one;
    BIG z, zilch;
    int bt;

    BIG_norm(b);
    BIG_copy(z, b);
    FP2_copy(&w, a);
    FP_one(&one);
    BIG_zero(zilch);
    FP2_from_FP(r, &one);
    while (1)
    {
        bt = BIG_parity(z);
        BIG_shr(z, 1);
        if (bt) FP2_mul(r, r, &w);
        if (BIG_comp(z, zilch) == 0) break;
        FP2_sqr(&w, &w);
    }
    FP2_reduce(r);
} */

/* test for x a QR */
int YYY::FP2_qr(FP2 *x,FP *h)
{ /* test x^(p^2-1)/2 = 1 */

    FP2 c;
    FP2_conj(&c,x);
    FP2_mul(&c,&c,x);

    return FP_qr(&(c.a),h);
}

/* sqrt(a+ib) = sqrt(a+sqrt(a*a-n*b*b))/2+ib/(2*sqrt(a+sqrt(a*a-n*b*b))/2) */

void YYY::FP2_sqrt(FP2 *w, FP2 *u, FP *h)
{
    FP w1, w2, w3, w4, hint;
    FP2 nw;
    int sgn,qr;
    FP2_copy(w, u);
    if (FP2_iszilch(w)) return;

    FP_sqr(&w1, &(w->b));  // b^2
    FP_sqr(&w2, &(w->a));  // a^2
    FP_add(&w1, &w1, &w2); FP_norm(&w1);  // a^2+b^2

    FP_sqrt(&w1, &w1, h);              // sqrt(a^2+b^2)  - could use an input hint to avoid exp!

    FP_add(&w2, &(w->a), &w1);            // a+sqrt(a^2+b^2)
    FP_norm(&w2);
    FP_div2(&w2, &w2);                    // w2=(a+sqrt(a^2+b^2))/2
// **
    FP_div2(&w1,&(w->b));                   // w1=b/2
    qr=FP_qr(&w2,&hint);                    // only exp!

// tweak hint
    FP_neg(&w3,&hint); FP_norm(&w3);    // QNR = -1
    FP_neg(&w4,&w2); FP_norm(&w4);

    FP_cmove(&w2,&w4,1-qr);
    FP_cmove(&hint,&w3,1-qr);

    FP_sqrt(&(w->a),&w2,&hint);             // a=sqrt(w2)
    FP_inv(&w3,&w2,&hint);                  // w3=1/w2
    FP_mul(&w3,&w3,&(w->a));                // w3=1/sqrt(w2)
    FP_mul(&(w->b),&w3,&w1);                // b=(b/2)*1/sqrt(w2)
    FP_copy(&w4,&(w->a));

    FP_cmove(&(w->a),&(w->b),1-qr);
    FP_cmove(&(w->b),&w4,1-qr);


/*

    FP_sqrt(&(w->a),&w2,&hint);             // a=sqrt(w2)
    FP_inv(&w3,&w2,&hint);                  // w3=1/w2
    FP_mul(&w3,&w3,&(w->a));                // w3=1/sqrt(w2)
    FP_mul(&(w->b),&w3,&w1);                // b=(b/2)*1/sqrt(w2)

// tweak hint
    FP_neg(&hint,&hint); FP_norm(&hint);    // QNR = -1
    FP_neg(&w2,&w2); FP_norm(&w2);

    FP_sqrt(&w4,&w2,&hint);                 // w4=sqrt(w2)
    FP_inv(&w3,&w2,&hint);                  // w3=1/w2    
    FP_mul(&w3,&w3,&w4);                    // w3=1/sqrt(w2)
    FP_mul(&w3,&w3,&w1);                    // w3=(b/2)*1/sqrt(w2)

    FP_cmove(&(w->a),&w3,1-qr);
    FP_cmove(&(w->b),&w4,1-qr);
*/
// return +ve root
    sgn=FP2_sign(w);
    FP2_neg(&nw,w); FP2_norm(&nw);
    FP2_cmove(w,&nw,sgn);
}

void YYY::FP2_rand(FP2 *x,csprng *rng)
{
    FP_rand(&(x->a),rng);
    FP_rand(&(x->b),rng);
}

File: javascript/src/fp2.js
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* Finite Field arithmetic  Fp^2 functions */

/* FP2 elements are of the form a+ib, where i is sqrt(-1) */

var FP2 = function(ctx) {
    "use strict";

    /* general purpose constructor */
    var FP2 = function(c, d) {
        if (c instanceof FP2) {
            this.a = new ctx.FP(c.a);
            this.b = new ctx.FP(c.b);
        } else {
            this.a = new ctx.FP(c);
            this.b = new ctx.FP(d);
        }
    };

    FP2.prototype = {
        /* reduce components mod Modulus */
        reduce: function() {
            this.a.reduce();
            this.b.reduce();
        },

        /* normalise components of w */
        norm: function() {
            this.a.norm();
            this.b.norm();
        },

        /* test this=0 ? */
        iszilch: function() {
            return (this.a.iszilch() && this.b.iszilch());
        },

        islarger: function() {
            if (this.iszilch()) return 0;
            var cmp=this.b.islarger();
            if (cmp!=0) return cmp;
            return this.a.islarger();
        },
        
        toBytes: function(bf) {
            var t = [];
            this.b.toBytes(t);
            for (var i=0;i<ctx.BIG.MODBYTES;i++)
                bf[i]=t[i];
            this.a.toBytes(t);
            for (var i=0;i<ctx.BIG.MODBYTES;i++)
                bf[i+ctx.BIG.MODBYTES]=t[i];            
        },

        /* test this=1 ? */
        isunity: function() {
            var one = new ctx.FP(1);
            return (this.a.equals(one) && this.b.iszilch());
        },

        /* conditional copy of g to this depending on d */
        cmove: function(g, d) {
            this.a.cmove(g.a, d);
            this.b.cmove(g.b, d);
        },

        /* test this=x */
        equals: function(x) {
            return (this.a.equals(x.a) && this.b.equals(x.b));
        },

        /* extract a */

		geta: function() {
			return this.a;
		},

        getA: function() {
            return this.a.redc();
        },

        /* extract b */
		getb: function() {
			return this.b;
		},

        getB: function() {
            return this.b.redc();
        },

        /* set from pair of FPs */
        set: function(c, d) {
            this.a.copy(c);
            this.b.copy(d);
        },

        /* set a */
        seta: function(c) {
            this.a.copy(c);
            this.b.zero();
        },

        /* set from two BIGs */
        bset: function(c, d) {
            this.a.bcopy(c);
            this.b.bcopy(d);
        },

        /* set from one ctx.BIG */
        bseta: function(c) {
            this.a.bcopy(c);
            this.b.zero();
        },

        /* copy this=x */
        copy: function(x) {
            this.a.copy(x.a);
            this.b.copy(x.b);
        },

        /* set this=0 */
        zero: function() {
            this.a.zero();
            this.b.zero();
        },

        /* set this=1 */
        one: function() {
            this.a.one();
            this.b.zero();
        },

        sign: function() {
            var p1=this.a.sign();
            var p2=this.b.sign();
            if (ctx.FP.BIG_ENDIAN_SIGN)
            {
                var u=this.b.iszilch()? 1:0;
                p2^=(p1^p2)&u;
                return p2;
            } else {
                var u=this.a.iszilch()? 1:0;
                p1^=(p1^p2)&u;
                return p1;
            }
        },

        /* negate this */
        neg: function() {
            var m = new ctx.FP(this.a),
                t = new ctx.FP(0);

            m.add(this.b);
            m.neg();
            t.copy(m);
            t.add(this.b);
            this.b.copy(m);
            this.b.add(this.a);
            this.a.copy(t);
        },

        /* conjugate this */
        conj: function() {
            this.b.neg();
            this.b.norm();
        },

        /* this+=a */
        add: function(x) {
            this.a.add(x.a);
            this.b.add(x.b);
        },

        /* this-=x */
        sub: function(x) {
            var m = new FP2(x); 
            m.neg();
            this.add(m);
        },

        rsub: function(x) {
            this.neg();
            this.add(x);
        },

        /* this*=s, where s is FP */
        pmul: function(s) {
            this.a.mul(s);
            this.b.mul(s);
        },

        /* this*=c, where s is int */
        imul: function(c) {
            this.a.imul(c);
            this.b.imul(c);
        },

        /* this*=this */
        sqr: function() {
            var w1 = new ctx.FP(this.a),
                w3 = new ctx.FP(this.a),
                mb = new ctx.FP(this.b);

            w1.add(this.b);
            w3.add(this.a);
            w3.norm();
            this.b.mul(w3);

            mb.neg();
            this.a.add(mb);

            this.a.norm();
            w1.norm();

            this.a.mul(w1);
        },

        /* this*=y */
        /* Now using Lazy reduction - inputs must be normed */
        mul: function(y) {
            var p = new ctx.BIG(0),
                pR = new ctx.DBIG(0),
                A, B, C, D, E, F;

            p.rcopy(ctx.ROM_FIELD.Modulus);
            pR.ucopy(p);

            if ((this.a.XES + this.b.XES) * (y.a.XES + y.b.XES) > ctx.FP.FEXCESS) {
                if (this.a.XES > 1) {
                    this.a.reduce();
                }

                if (this.b.XES > 1) {
                    this.b.reduce();
                }
            }

            A = ctx.BIG.mul(this.a.f, y.a.f);
            B = ctx.BIG.mul(this.b.f, y.b.f);

            C = new ctx.BIG(this.a.f);
            D = new ctx.BIG(y.a.f);

            C.add(this.b.f);
            C.norm();
            D.add(y.b.f);
            D.norm();

            E = ctx.BIG.mul(C, D);
            F = new ctx.DBIG(0);
            F.copy(A);
            F.add(B);
            B.rsub(pR);

            A.add(B);
            A.norm();
            E.sub(F);
            E.norm();

            this.a.f.copy(ctx.FP.mod(A));
            this.a.XES = 3;
            this.b.f.copy(ctx.FP.mod(E));
            this.b.XES = 2;
        },

        /* this^e */
/*
        pow: function(e) {
            var r = new FP2(1),
                w = new FP2(this), 
                z = new ctx.BIG(e),
                bt;

            for (;;) {
                bt = z.parity();
                z.fshr(1);
                if (bt == 1) {
                    r.mul(w);
                }

                if (z.iszilch()) {
                    break;
                }
                w.sqr();
            }

            r.reduce();
            this.copy(r);
        }, */

        qr: function(h) {
            var c = new FP2(this);
            c.conj();
            c.mul(this);

            return c.geta().qr(h);
        },

        /* sqrt(a+ib) = sqrt(a+sqrt(a*a-n*b*b)/2)+ib/(2*sqrt(a+sqrt(a*a-n*b*b)/2)) */
        sqrt: function(h) {
            var w1, w2, w3, w4, hint;

            if (this.iszilch()) {
                return;
            }

            w1 = new ctx.FP(this.b);
            w2 = new ctx.FP(this.a);
            w3 = new ctx.FP(this.a);
            w4 = new ctx.FP(this.a);
            hint = new ctx.FP(this.a);

            w1.sqr();
            w2.sqr();
            w1.add(w2); w1.norm();

            w1 = w1.sqrt(h);

            w2.copy(this.a);
            w2.add(w1);
            w2.norm();
            w2.div2();

            w1.copy(this.b); w1.div2();
            var qr=w2.qr(hint);

// tweak hint
            w3.copy(hint); w3.neg(); w3.norm();
            w4.copy(w2); w4.neg(); w4.norm();

            w2.cmove(w4,1-qr);
            hint.cmove(w3,1-qr);

            this.a.copy(w2.sqrt(hint));
            w3.copy(w2); w3.inverse(hint);
            w3.mul(this.a);
            this.b.copy(w3); this.b.mul(w1);
            w4.copy(this.a);

            this.a.cmove(this.b,1-qr);
            this.b.cmove(w4,1-qr);

/*
            this.a.copy(w2.sqrt(hint));
            w3.copy(w2); w3.inverse(hint);
            w3.mul(this.a);
            this.b.copy(w3); this.b.mul(w1);

            hint.neg(); hint.norm();
            w2.neg(); w2.norm();

            w4.copy(w2.sqrt(hint));
            w3.copy(w2); w3.inverse(hint);
            w3.mul(w4);
            w3.mul(w1);

            this.a.cmove(w3,1-qr);
            this.b.cmove(w4,1-qr);
*/
            var sgn=this.sign();
            var nr=new FP2(this);
            nr.neg(); nr.norm();
            this.cmove(nr,sgn);
        },

        /* convert this to hex string */
        toString: function() {
            return ("[" + this.a.toString() + "," + this.b.toString() + "]");
        },

        /* this=1/this */
        inverse: function(h) {
            var w1, w2;

            this.norm();

            w1 = new ctx.FP(this.a);
            w2 = new ctx.FP(this.b);

            w1.sqr();
            w2.sqr();
            w1.add(w2);
            w1.inverse(h);
            this.a.mul(w1);
            w1.neg();
            w1.norm();
            this.b.mul(w1);
        },

        /* this/=2 */
        div2: function() {
            this.a.div2();
            this.b.div2();
        },

        /* this*=sqrt(-1) */
        times_i: function() {
            var z = new ctx.FP(this.a); //z.copy(this.a);
            this.a.copy(this.b);
            this.a.neg();
            this.b.copy(z);
        },

        /* w*=(1+sqrt(-1)) */
        /* where X*2-(1+sqrt(-1)) is irreducible for FP4, assumes p=3 mod 8 */
        mul_ip: function() {
			var t = new FP2(this);
			var i = ctx.FP.QNRI;
			this.times_i();
			while (i>0)
			{
				t.add(t);
				t.norm();
				i--;
			}
			this.add(t);
			if (ctx.FP.TOWER==ctx.FP.POSITOWER) {
				this.norm();
				this.neg();
			}
        },

        /* w/=(1+sqrt(-1)) */
        div_ip: function() {
			var z=new FP2(1<<ctx.FP.QNRI,1);
			z.inverse(null);
			this.mul(z);
			if (ctx.FP.TOWER==ctx.FP.POSITOWER) {
				this.neg();
				this.norm();
			}
        }
    };

    FP2.rand = function(rng) {
        return new FP2(ctx.FP.rand(rng),ctx.FP.rand(rng));
    };

    FP2.fromBytes = function(bf) {
        var t=[];
        for (var i=0;i<ctx.BIG.MODBYTES;i++)
            t[i]=bf[i];
        var tb=ctx.FP.fromBytes(t);
        for (var i=0;i<ctx.BIG.MODBYTES;i++)
            t[i]=bf[i+ctx.BIG.MODBYTES];
        var ta=ctx.FP.fromBytes(t);
        return new FP2(ta,tb);
    };

    return FP2;
};

if (typeof module !== "undefined" && typeof module.exports !== "undefined") {
    module.exports = {
        FP2: FP2
    };
}

